### DISCUSSION TITLE
Low Power sketch ATtiny861V - need help

### DISCUSSION THREAD

**Discussion by jzobac** [SpenceKonde/ATTinyCore#638] [General]
Hi,
I have try some magic, but get only 620uA in normal operation (datasheet says 300uA) and 3.7uA in sleep (datasheet says 0.1uA).

Fuse configuration I will write later (I'm not at home).

And if someone known how to internally read VCC, it would be great too.

```
#include <avr/power.h>
#include <avr/wdt.h>
#include <avr/sleep.h>

float vcc;


float readVcc() {
   power_adc_enable();
    // Read 1.1V reference against AVcc
    // set the reference to Vcc and the measurement to the internal 1.1V reference
#if defined(__AVR_ATmega32U4__) || defined(__AVR_ATmega1280__) || defined(__AVR_ATmega2560__)
    ADMUX = _BV(REFS0) | _BV(MUX4) | _BV(MUX3) | _BV(MUX2) | _BV(MUX1);
#elif defined (__AVR_ATtiny24__) || defined(__AVR_ATtiny44__) || defined(__AVR_ATtiny84__)
    ADMUX = _BV(MUX5) | _BV(MUX0);
#elif defined (__AVR_ATtiny25__) || defined(__AVR_ATtiny45__) || defined(__AVR_ATtiny85__)
    ADMUX = _BV(MUX3) | _BV(MUX2);
#else
    ADMUX = _BV(REFS0) | _BV(MUX3) | _BV(MUX2) | _BV(MUX1);
#endif

    delay(2); // Wait for Vref to settle
    ADCSRA |= _BV(ADSC); // Start conversion
    
    while (bit_is_set(ADCSRA,ADSC)); // measuring
     power_adc_disable();
    //unsigned int a = ADCW;
    unsigned int result = 1119162UL / ADCW; // Calculate Vcc (in mV)
    return result / 1000.0; // Vcc in volts*10
}

// Turn on watchdog timer; interrupt mode every 8.0s
void watchdogEnable8s()
{
    cli();
    MCUSR = 0;
    WDTCR |= B00011000;                  //watchdog configuration mode
    //WDTCR = B01000100;                // 250 miliSecond Timeout
    //WDTCR = B01000111;                // 2 Second Timeout
    //WDTCR = B01100000;                // 4 Second Timeout
    WDTCR = B01100001;                  // 8 Second Timeout
    sei();
}

void sleepNow()
{
    byte adcsra_save = ADCSRA;                // save the ADC Control and Status Register A
    //byte prr_save = PRR; //pak to nemeri napeti
    ADCSRA = 0;                          // disable the ADC (saves 300 µA)
    //PRR = 0xFF; // turn off various modules
    set_sleep_mode(SLEEP_MODE_PWR_DOWN); // Sleep mode is set here
    cli();
    sleep_enable();                      // Enables the sleep bit in the mcucr register
                                         // so sleep is possible. just a safety pin
    //sleep_bod_disable();                 // disable brown-out detection while sleeping (20-25 µA)
    sei();
    sleep_cpu();                         // Here the device is actually put to sleep!!
                                         // THE PROGRAM CONTINUES FROM HERE AFTER WAKING UP
    sleep_disable();                     // First thing after waking from sleep:
                                         // disable sleep...
    ADCSRA = adcsra_save;                     // restore ADCSRA = re-enable the ADC
    //PRR = prr_save;
}

// WDT Wakeup
ISR(WDT_vect)
{
    wdt_disable();
}




void setup()
{
  //power_adc_disable(); // ADC converter
  //power_spi_disable(); // SPI
  //power_usart0_disable();// Serial (USART)
  //power_timer0_disable();// Timer 0
  power_timer1_disable();// Timer 1
//  power_timer2_disable();// Timer 2
//  power_twi_disable(); // TWI (I2C)


//  vcc = readVcc();    //not migrated from atmega328p yet
  
  power_adc_disable(); // ADC converter
}

void loop()
{
  
 
  power_adc_enable(); // ADC converter  (atmega has lower power consumption with this in sleep; will try without this)

        wdt_reset();                         // Get ready to go to sleep...
        watchdogEnable8s();                    // Turn on the watchdog timer
        sleepNow();                          // Go to sleep, watchdog timer will wake later

power_adc_disable(); // ADC converter

delay(8000);
  
}

```

jzobac

**Comment by SpenceKonde** [comment]
You are factoring in the expected power consumption of the WDT there? The other thing - make sure all pins are in a defined state, no floating pins. Transitions on the input buffers can use a small amount of power. 

I also notice you're leaving timer0 enabled, is that intended? For minimum power usage, you'd want to turn off power to timer0 (disable millis from tools menu too in this case)
You also probably want to disable the power to the USI (that's the crapola substitute that these parts have for TWI and SPI, which is shittier by far than either of those two, and barely better than bitbanging. 

The 861 uses the 4 least significant bits of the PRR:  0x08 is timer1 which I see you are shutting off. 0x04 is timer0, which you are not, 0x02 is the USI which you are not, and 0x01 is the ADC. I reckon you'd want to disable millis from the menu option, and set `PRR = 0x0E (that's 0b00001110 - timer1, timer0, and USI power disconnected, and ADC power connected)` in setup. And before sleep, disable the ADC, and set `PRR = 0x0F` (everything power disconnected) and then set `PRR = 0x0E` again before enabling the ADC upon wake.

**Reply by jzobac** [reply]
Hi,
thanks, will try. I have migrated sketch from atmega328p, not all is yet done. I doesn't known if I will need millis or not, I will (wish to) use software serial and I2C.

Here are my fuses
avrdude: safemode: lfuse reads as 62
avrdude: safemode: hfuse reads as D7
avrdude: safemode: efuse reads as FF

http://eleccelerator.com/fusecalc/fusecalc.php?chip=attiny861&LOW=62&HIGH=DF&EXTENDED=FF&LOCKBIT=FF

BOD disabled, 1Mhz.

jzobac

**Comment by SpenceKonde** [comment]
One thought.... ah, how do I say this.... Why are you using rather old 861V
with it's bizarre timers, unique limitations and decade-old design, when
you're not using any of it's unique features (mostly at this point, the
high speed timer and 3-phase BLDC/bridge control capabilities, which are
still unmatched with no sign of that changing) ... and you need serial and
I2C on a part with  on a part with no hardware serial port and a terribad
USI peripheral instead of a real I2C interface? Do you have piles of these
kicking around or something?

Because like... you can check off all the features (including a
differential ADC with programmable gain, seen on a new AVR for the first
time in like 8 years, hardware I2C & SPI and not one but two hardware
serial ports) with none of the drawbacks, power usage at least as low, and
get 4 times the flash and 6 times the RAM for around the same price on a
20-pin tinyAVR available in SOIC20 with the recently released tinyAVR
2-series, ex 3226, 1626, 826, 426)? It seems like there's no real reason
for you to use the 861 here,,, Arduino support for the modern tinyAVRs is
available through my megaTinyCore. They've also got a pile of other cool
features, most of which have arduino library wrappers. That and DxCore are
the reason updates are so much less frequent on ATTinyCore now....

On Fri, Jan 7, 2022 at 6:23 AM jzobac ***@***.***> wrote:

> Hi,
> thanks, will try. I have migrated sketch from atmega328p, not all is yet
> done. I doesn't known if I will need millis or not, I will (wish to) use
> software serial and I2C.
>
> Here are my fuses
> avrdude: safemode: lfuse reads as 62
> avrdude: safemode: hfuse reads as D7
> avrdude: safemode: efuse reads as FF
>
>
> http://eleccelerator.com/fusecalc/fusecalc.php?chip=attiny861&LOW=62&HIGH=DF&EXTENDED=FF&LOCKBIT=FF
>
> BOD disabled, 1Mhz.
>
> jzobac
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/SpenceKonde/ATTinyCore/discussions/638#discussioncomment-1924246>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABTXEW7A25E33FYSU26OXNTUU3ELJANCNFSM5LMQVRGQ>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
> You are receiving this because you commented.Message ID:
> ***@***.***>
>


-- 
____________
Spence Konde
Azzy’S Electronics

New products! Check them out at tindie.com/stores/DrAzzy
GitHub: github.com/SpenceKonde
ATTinyCore <https://github.com/SpenceKonde/ATTinyCore>: Arduino support for
all pre-2016 tinyAVR with >2k flash!
megaTinyCore <https://github.com/SpenceKonde/megaTinyCore>: Arduino support
for all post-2016 tinyAVR parts!
DxCore <https://github.com/SpenceKonde/DxCore>: Arduino support for the AVR
Dx-series parts, the latest and greatest from Microchip!
Contact: ***@***.***

**Comment by jzobac** [comment]
Easy answer :), I'm not able to get them. Attiny85V is unavailable maybe 5 years (only batch of 1000+ is maybe). New Attiny's are with no chance to get them in THT. And you write on the github "don't buy them in china, most of them are fakes".

OK will revert back to Arduino Pro Mini - power usage seems to be little more than of an ATtiny861V (and in TME they have listed that has I2C)
https://www.tme.eu/cz/details/attiny861v-10pu/rodina-avr-8-bitu/microchip-atmel/

https://www.tme.eu/cz/katalog/mikrokontrolery-a-mikroprocesory_100590/?s_field=1000014&s_order=asc&limit=20&currency=CZK&visible_params=10%2C98%2C2%2C367%2C788%2C35%2C2826%2C375%2C804%2C790%2C789%2C783%2C32%2C793%2C409%2C2911%2C787%2C2946%2C2912%2C2408%2C403%2C792%2C794%2C2910%2C1481%2C31%2C2411%2C909%2C798%2C1600%2C193%2C797%2C2947%2C2914&mapped_params=10%3A1436576%3B98%3A1474124%2C1444190%3B783%3A1453431%2C1453439%2C1453452%2C1453462%2C1502815%3B&gross_prices=1&page=1

same in gme
https://www.gme.cz/mikroprocesory-avr?onpage=80#products

If you will send from EU, I will buy batch of 20 Tinny's. And if they prove that are good and I can work with them, than more. Customs in czech are terrible.

jzobac

**Comment by hmeijdam** [comment]
Mouser seems to have a few AT85V [in stock.](https://nl.mouser.com/c/semiconductors/embedded-processors-controllers/microcontrollers-mcu/8-bit-microcontrollers-mcu/?q=attiny85v&instock=y)

**Comment by jzobac** [comment]
I have contacted Mouser with question - from where they ship from and after 4 days no response. For me not good partner to buy something.

Have ordered Attiny3224 (from TME), but when I look to datasheet, the power consumption is +- equal to Atmega328p, I don't see benefits of new attinys. See just one, 2x HW UART, that's for me useful, one for application and second for debug. But this is only useful for development, thus for production use is Atmega328p good enough (and for development is enough sw serial too).
